ext/drsyms/drsyms_unix_common.c

```diff
@@ -439,6 +460,114 @@ addrsearch_symtab(dbg_module_t *mod, size_t modoffs, drsym_info_t *info INOUT, u
+ * than strlen() will find.
+ * XXX: Should DR export a malloc-matching heap API that does not start with
+ * underscores for cleaner usage?  Most non-static-DR usage though can just
+ * use malloc().  Here we want to support static DR.
+ *
+ * We do not want to pay the cost of hashtable_t doing a further
+ * strdup on our just-allocated strings so we free keys ourselves.
+ */
+
+static void
+drsym_free_hash_key(void *key)
+{
+    __wrap_free(key);
+}
+
+static char *
+drsym_dup_string_until_char(const char *sym, char stop)
+{
+    /* We skip the first char to avoid empty strings. */
+    const char *found = strchr(sym + 1, stop);
+    if (found != NULL) {
+        size_t no_found_sz = found - sym;
+        char *no_found = (char *)__wrap_malloc(no_found_sz + 1);
+        dr_snprintf(no_found, no_found_sz, "%s", sym);
+        no_found[no_found_sz] = '\0';
+        return no_found;
+    }
+    return NULL;
+}
+
+static char *
+drsym_demangle_helper(const char *sym, uint flags)
+{
+    /* We look for "_Z" to avoid the copy of the same symbol for non-mangled cases. */
+    if (sym[0] != '_' || sym[1] != 'Z')
+        return NULL;
+    size_t len;
+    size_t buf_size = 1024;
+    char *buf = (char *)__wrap_malloc(buf_size);
+    /* Resize until it's big enough. */
+    while ((len = drsym_demangle_symbol(buf, buf_size, sym, flags)) > buf_size) {
+        __wrap_free(buf);
+        buf_size = len;
+        buf = (char *)__wrap_malloc(buf_size);
+    }
+    if (len <= 0) {
+        return NULL;
+    }
+    return buf;
+}
+
+static bool
+drsym_add_hash_entry(dbg_module_t *mod, const char *copy, size_t modoffs)
+{
+    if (hashtable_add(&mod->symtable, (void *)copy, (void *)modoffs)) {
+        NOTIFY("%s: added %s\n", __FUNCTION__, copy);
+        return true;
+    }
+    drsym_free_hash_key((void *)copy);
+    return false;
+}
+
+/* Symbol enumeration callback for doing a single lookup.
+ */
+static bool
+drsym_fill_symtable_cb(const char *sym, size_t modoffs, void *data INOUT)
+{
+    dbg_module_t *mod = (dbg_module_t *)data;
+    size_t len = strlen(sym);
+    char *copy = __wrap_malloc(len + 1);
+    dr_snprintf(copy, len, "%s", sym);
+    copy[len] = '\0';
+    if (!drsym_add_hash_entry(mod, copy, modoffs))
+        return true;
+
+    /* We match the name part of versioned symbols: so "foo" will match
+     * "foo@@GLIBC_2.1".  If there are multiple, a user who wants one in
+     * particular needs to include the version name in the search target.
+     */
+    char *toadd = drsym_dup_string_until_char(sym, '@');
+    if (toadd != NULL)
+        drsym_add_hash_entry(mod, toadd, modoffs);
+
+    /* Add the demanglings. */
+    toadd = drsym_demangle_helper(sym, DRSYM_DEMANGLE);
+    if (toadd == NULL)
+        return true;
+    if (!drsym_add_hash_entry(mod, toadd, modoffs))
+        return true;
+
+    /* Add a version without parameters to allow the user to ignore overloads. */
+    char *noparen = drsym_dup_string_until_char(toadd, '(');
```

Comment:
Unfortunately, C++ has a lot of cases where the simple '(' heurstic doesn't work, like `Foo::(anonymous namespace)::bar()`, or `std::function<int(int)>::foo()`.
[Here](https://github.com/google/pprof/blob/be90f332299a1d03fbb09bdbebdf28118ae300fb/profile/prune_test.go#L55)'s a few examples.

Reply:
Fair, but the initial goal was to match the existing code which just looked at the first paren.  I'll add a comment about the problems with this.